File: packages/shared-types/src/position-metrics.ts
export interface FlightPositionMetricsInput {
  altitudeHundredsFeet: number | null | undefined;
  groundspeedKnots: number | null | undefined;
  headingDegrees: number | null | undefined;
}

export interface FlightPositionMetrics {
  altitudeHundredsFeet: number | null;
  altitudeFeet: number | null;
  groundspeedKnots: number | null;
  headingDegrees: number | null;
}

export function normalizeAltitudeFeet(
  altitudeHundredsFeet: number | null | undefined,
): number | null {
  return altitudeHundredsFeet === null || altitudeHundredsFeet === undefined
    ? null
    : altitudeHundredsFeet * 100;
}

export function normalizeGroundspeedKnots(
  groundspeedKnots: number | null | undefined,
): number | null {
  return groundspeedKnots ?? null;
}

export function normalizeHeadingDegrees(headingDegrees: number | null | undefined): number | null {
  if (headingDegrees === null || headingDegrees === undefined) {
    return null;
  }

  return headingDegrees === 360 ? 0 : headingDegrees;
}

export function normalizeFlightPositionMetrics(
  input: FlightPositionMetricsInput,
): FlightPositionMetrics {
  const altitudeHundredsFeet = input.altitudeHundredsFeet ?? null;

  return {
    altitudeHundredsFeet,
    altitudeFeet: normalizeAltitudeFeet(altitudeHundredsFeet),
    groundspeedKnots: normalizeGroundspeedKnots(input.groundspeedKnots),
    headingDegrees: normalizeHeadingDegrees(input.headingDegrees),
  };
}

File: services/internal/domain/position_metrics.go
package domain

type FlightPositionMetricsInput struct {
	AltitudeHundredsFeet *int
	GroundspeedKnots     *int
	HeadingDegrees       *int
}

type FlightPositionMetrics struct {
	AltitudeHundredsFeet *int
	AltitudeFeet         *int
	GroundspeedKnots     *int
	HeadingDegrees       *int
}

func NormalizeAltitudeFeet(altitudeHundredsFeet *int) *int {
	if altitudeHundredsFeet == nil {
		return nil
	}

	altitudeFeet := *altitudeHundredsFeet * 100
	return &altitudeFeet
}

func NormalizeGroundspeedKnots(groundspeedKnots *int) *int {
	return groundspeedKnots
}

func NormalizeHeadingDegrees(headingDegrees *int) *int {
	if headingDegrees == nil {
		return nil
	}

	normalized := *headingDegrees
	if normalized == 360 {
		normalized = 0
	}

	return &normalized
}

func NormalizeFlightPositionMetrics(input FlightPositionMetricsInput) FlightPositionMetrics {
	return FlightPositionMetrics{
		AltitudeHundredsFeet: input.AltitudeHundredsFeet,
		AltitudeFeet:         NormalizeAltitudeFeet(input.AltitudeHundredsFeet),
		GroundspeedKnots:     NormalizeGroundspeedKnots(input.GroundspeedKnots),
		HeadingDegrees:       NormalizeHeadingDegrees(input.HeadingDegrees),
	}
}
